pallene/assignment_conversion.lua

```diff
@@ -0,0 +1,347 @@
+local util     = require "pallene.util"
+local typedecl = require "pallene.typedecl"
+local types    = require "pallene.types"
+local ast      = require "pallene.ast"
+
+local converter = {}
+
+--
+--  ASSIGNMENT CONVERSION
+-- ======================
+-- This Compiler pass handles mutable captured variables inside nested closures by
+-- transforming AST Nodes that reference or re-assign to them. All captured variables
+-- are 'boxed' inside records. For example, consider this code:
+-- ```
+-- function m.foo()
+--     local x: integer = 10
+--     local set_x: (integer) -> () = function (y)
+--          x = y
+--     end
+--     local get_x: () -> integer = function () return x end
+-- end
+--```
+-- The AST representation of the above snippet, will be converted in this pass to
+-- something like the following:
+-- ```
+-- record $T
+--     value: integer
+-- end
+--
+-- function m.foo()
+--     local x: $T = { value = 10 }
+--     local set_x: (integer) -> () = function (y)
+--          x.value = y
+--     end
+--     local get_x: () -> integer = function () return x.value end
+-- end
+-- ```
+
+local Converter = util.Class()
+
+function converter.convert(prog_ast)
+    return Converter.new():visit_prog(prog_ast)
+end
+
+
+local function FuncInfo()
+    return {
+        mutated_decls  = {}, -- list of ast.Decl
+        captured_decls = {} -- { ast.Decl }
+    }
+end
+
+function Converter:init()
+    self.func_stack = {} -- list of FuncInfo
+    local func_info = FuncInfo()
+    table.insert(self.func_stack, func_info)
+
+    self.ref_to_decl        = {} -- list of ast.Var.Name
+    self.func_depth_of_decl = {} -- { ast.Decl.Decl => integer }
+    self.init_exp_of_decl   = {} -- { ast.Decl.Decl => ast.Exp }
+    self.box_records        = {} -- list of ast.Toplevel.Record
+
+    -- used to assign unique names to subsequently generated record types
+    self.typ_counter = 0
+end
+
+-- generates a unique type name each time
+function Converter:type_name(var_name)
+    self.typ_counter = self.typ_counter + 1
+    return "$T_"..var_name.."_"..self.typ_counter
+end
+
+function Converter:add_box_type(loc, typ)
+    local dummy_node = ast.Toplevel.Record(loc, types.tostring(typ), {})
+    dummy_node._type = typ
+    table.insert(self.box_records, dummy_node)
+end
+
+function Converter:register_decl(decl)
+    if not self.ref_to_decl[decl] then
+        self.ref_to_decl[decl]        = {}
+        self.func_depth_of_decl[decl] = #self.func_stack
+    end 
+end
+
+function Converter:visit_prog(prog_ast)
+    assert(prog_ast._tag == "ast.Program.Program")
+    for _, tl_node in ipairs(prog_ast.tls) do
+        if tl_node._tag == "ast.Toplevel.Stats" then
+            for _, stat in ipairs(tl_node.stats) do
+                self:visit_stat(stat)
+            end
+        else
+            -- skip record declarations and type aliases
+            assert(typedecl.match_tag(tl_node._tag, "ast.Toplevel"))
```

Comment:
Is this else case really necessary?

Reply:
I just added it to make sure that we're not ever calling this function with the wrong argument.

Reply:
Okk